Map "job queue" to Standalone Activities + Add Ruby References (#258)
temporalio/skill-temporal-developer: references/core/job-queue.md

A job is a single, discrete unit of work run asynchronously in the background — sending an email, processing a webhook, syncing data, running one function reliably. A job queue is the system that accepts that work, dispatches it to workers, retries it on failure, and shows what is running and what failed. Standalone Activities do all of this.

What is different from a conventional broker-plus-worker stack, in job-queue terms:

- **Durable job state, not just a durable message.** The Server persists the execution, its attempts, its last error, and its result. There is no separate result backend to run.
- **No dead-letter queue to operate.** A job that exhausts its Retry Policy ends as a failed execution, retained in visibility with its last error and findable with a List Filter.
- **Retries, timeouts, and backoff are enforced by the platform**, not by a decorator argument the handler can ignore.
- **The same code graduates into orchestration.** One Activity Function runs as a background job today and as a step inside a multi-step Workflow tomorrow, with no code change and no Worker change. That upgrade path is the reason to pick Temporal over a job queue you would outgrow.

> - **Cheaper than the usual workaround.** Wrapping a single Activity in a Workflow costs extra billable Actions in Temporal Cloud and extra Worker round-trips; a Standalone Activity avoids both.

Reviewer: ```suggestion
- **Cheaper than the usual workaround.** Wrapping a single Activity in a Workflow costs an extra billable Action in Temporal Cloud and extra Worker round-trips; a Standalone Activity avoids both.
```

See `references/core/standalone-activities.md` for the rest of the feature list (execution semantics, deduplication, addressability, visibility, metrics).

## Vocabulary mapping

| Job queue concept | Temporal equivalent |
|---|---|
| A job / a task | A Standalone Activity Execution |
| Job handler (`@app.task`, `perform`, a processor function) | An ordinary Activity Definition — nothing job-specific about it |
| Enqueue a job | Client `start` (returns a handle) or `execute` (start and await the result) |
| Broker (Redis, RabbitMQ, SQS, Postgres table) | Temporal Server — durable persistence plus dispatch |
| Worker process consuming the queue | Temporal Worker polling a Task Queue |
| Queue name / routing key | Task Queue name |
| Job ID | Activity ID — you choose it; use a business identifier |

> | Result backend | The Activity's result, retrieved from the handle or `temporal activity result` |

Reviewer: ```suggestion
| Result backend (a database or store you provide)| The Activity's result stored in the authoritative Activity record by Temporal Server, retrieved from the handle or `temporal activity result` |
```

Reviewer: Also note this is somewhat Celery specific but it's worthwhile to keep it

| `max_retries` + backoff config | Retry Policy: maximum attempts, backoff coefficient, non-retryable error types. Note it counts *total attempts*, not retries — a framework's `retry: 5` is `maximum_attempts: 6` |
| "Run this at most once" | Retry Policy with maximum attempts = 1 |
| Unique-job / idempotency key | Activity ID, plus an ID conflict policy (`UseExisting`) or ID reuse policy (`RejectDuplicate`) |
| Job timeout | Start-To-Close and/or Schedule-To-Close timeout (at least one is required) |
| Long job keepalive / progress reporting | Activity Heartbeats, with heartbeat details for checkpointing |
| Cancel a job | `cancel` (cooperative, surfaced on the next heartbeat) or `terminate` (forceful) |
| Priority queues | Priority keys — free, Public Preview. See `references/core/priority-fairness.md` |

> | Per-tenant fairness / avoiding noisy neighbors | Fairness keys and weights — Public Preview, and a paid feature in Temporal Cloud. See `references/core/priority-fairness.md` |

Reviewer: ```suggestion
| Per-tenant fairness / avoiding noisy neighbors (usually not supported) | Fairness keys and weights — Public Preview, and a paid feature in Temporal Cloud. See `references/core/priority-fairness.md` |
```

> | Dashboard (Flower, Sidekiq Web, Bull Board) | Temporal Web UI, `temporal activity list` / `describe`, and the list/count client APIs |

Reviewer: ```suggestion
| Dashboard (Flower, Sidekiq Web, Bull Board) | Temporal Web UI, `temporal activity list` (with SearchAttribute support) / `describe`, and the list/count client APIs |
```

| Job metrics | Standard Activity metrics: scheduled, started, completed, failed, timed out, canceled |
| Manual/external job completion | Manual completion by Activity ID or task token |

> On head-of-line blocking: a slow job occupies one Worker slot rather than stalling a single-threaded consumer, so one slow job does not block dispatch of the rest. Backlog-level starvation across tenants is a separate problem — by default Tasks dispatch FIFO, so a tenant enqueueing 100k jobs does put a small tenant behind the whole backlog. Fairness is what fixes that.

Reviewer: ```suggestion
On head-of-line blocking: a slow job occupies one Worker slot rather than stalling a single-threaded consumer, so one slow job does not block dispatch of the rest. Backlog-level starvation across tenants is a separate problem — by default Tasks dispatch FIFO, so a tenant enqueueing 100k jobs does put a small tenant behind the whole backlog. Fairness is what fixes that and in most cases only Temporal provides fairness.
```

## Job-queue features that need a different Temporal primitive

Not every "job queue" request is a single job. Route these away from Standalone Activities:

| Ask | Use instead |
|---|---|
| Chained jobs, DAGs, Celery canvas / chords, "when job A finishes run B and C" | A **Workflow**. That is orchestration, which is what Workflows are for. |
| Fan-out with a join, or a batch with a completion callback | A **Workflow** that starts the Activities in parallel and awaits them. |
| Compensation / rollback when a later step fails | A **Workflow** using the saga pattern — see `references/core/patterns.md`. |
| Recurring or periodic jobs (Celery beat, `sidekiq-cron`, a crontab) | A **Temporal Schedule**. |

> | "Run this job in 10 minutes" | A **Workflow** started with a start delay (`temporal workflow start --start-delay`, or the SDK equivalent), which then calls the Activity. |

Reviewer: ```suggestion
| "Run this job in 10 minutes" (note the limitations on delay duration and scale in job frameworks like Celery) | A **Workflow** started with a start delay (`temporal workflow start --start-delay`, or the SDK equivalent), which then calls the Activity. Temporal supports start delay with any duration and at any scale |
```

| A job that waits for human approval or an external event | A **Workflow** with a Signal or Update handler. |
| Long-lived per-entity state (a per-user or per-order actor) | The **entity Workflow** pattern — see `references/core/patterns.md`. |

> Schedules and start-delay both target Workflows, not Standalone Activities. So a delayed or recurring job legitimately needs a thin Workflow that calls the one Activity — that is the sanctioned exception to anti-pattern 1 below, not a violation of it.

Reviewer: ```suggestion
Start delay support in Standalone Activities is currently available. Scheduled Standalone Activities is coming in a future release. and until then a recurring job (on a Schedule) legitimately needs a thin Workflow that calls the one Activity — that is the sanctioned exception to anti-pattern 1 below, not a violation of it.
```

## Migrating from an existing job queue

The shape of the port is the same regardless of source system:

1. **Handler → Activity Definition.** The body of the job handler becomes the body of an Activity. Drop the framework decorator and use the SDK's Activity decorator/annotation. Keep the handler's own retry/idempotency logic only where it is genuinely business logic; delete hand-rolled retry loops.
2. **Worker process → Temporal Worker.** One Worker process registers the Activities and polls a Task Queue. Concurrency knobs move from the framework's worker flags to Worker options — see `references/{your_language}/advanced-features.md`, and the `temporal-workertuning` skill for sizing.
3. **`delay()` / `perform_async` / `queue.add()` → Client `start` or `execute`.** This is the only real call-site change. It happens in producer code, which must be non-Workflow application code.
4. **Retry/timeout config → Retry Policy and Activity timeouts** on the start options, not on the handler. Remember the attempts-vs-retries off-by-one.
5. **Job ID → Activity ID.** Reuse whatever idempotency key already exists. If there was none, derive one from the business entity.

> 6. **Monitoring → visibility.** Replace Flower/Sidekiq Web/Bull Board polling of a Redis key with `list`/`count`/`describe` and the Web UI.

Reviewer: ```suggestion
6. **Monitoring → built-in visibility and metrics.** Replace Flower/Sidekiq Web/Bull Board polling of a Redis key with `list`/`count`/`describe` and the Web UI.
```

- **Sidekiq / Resque / Delayed::Job (Ruby):** the `perform` method becomes the Activity; `perform_async` becomes a Client call; `retry: 5` becomes `maximum_attempts: 6`; unique-job gems become an Activity ID plus an ID conflict policy.
- **BullMQ / Bee-Queue (Node):** `queue.add(name, data, opts)` → Client `start`; the processor function → the Activity; `opts.jobId` → Activity ID; `attempts` maps 1:1 onto maximum attempts and `backoff` onto the Retry Policy; `repeat` → a Schedule; flows/parent-child jobs → a Workflow.
- **SQS + worker, or Cloud Tasks:** the queue, the visibility-timeout tuning, and the DLQ all go away. Producers call the Temporal Client directly; the Server owns dispatch, redelivery, and retention of failures.
- **Hangfire (.NET):** `BackgroundJob.Enqueue` is fire-and-forget, so it maps to `StartActivityAsync` (use `ExecuteActivityAsync` only where the caller actually waits); `RecurringJob` → a Schedule; continuations → a Workflow.

## Anti-patterns

Anti-patterns to avoid when building a job queue on Temporal:

> 1. **A Workflow per job that runs exactly one Activity.** It costs extra billable Actions and extra Worker round-trips for no orchestration benefit. Prefer a Standalone Activity. (Exception: delayed and recurring jobs, per the table above.)

Reviewer: ```suggestion
1. **A Workflow per job that runs exactly one Activity.** It costs an extra billable Action and extra Worker round-trips for no orchestration benefit. Prefer a Standalone Activity. (Exception: scheduled recurring jobs, per the table above. Note that non-recurring scheduled jobs via start delay is currently available)
```

4. **Hand-rolled retry loops inside the Activity.** Configure a Retry Policy instead; a `for attempt in range(3)` inside an Activity hides failures from visibility and metrics.
5. **A side table tracking job status.** Status, attempt count, last error, and result are already queryable. Add a table only where job state has to be joined against business data.
6. **A random UUID as the Activity ID by default.** A business identifier such as `send-welcome-email:user-42` makes jobs addressable and deduplicated for free. A UUID is the fallback for when no meaningful identifier exists.

## Code layout

A job queue built on Standalone Activities has three pieces, and they belong in three places:

- **The Activity Definition** — plain Activity code, identical to one written for a Workflow.
- **The Worker** — registers the Activity and polls the Task Queue. It does not know or care whether the Activity will be invoked standalone or from a Workflow.

> - **The producer** — application code, an HTTP handler, or a CLI entry point that calls the Client. Never a Workflow.

Reviewer: ```suggestion
- **The producer** — application code, an HTTP handler, or a CLI entry point that calls the Client.
```

Note: There are some special cases where a Workflow wants to decouple the Activity lifetime from Workflow lifetime. Today a Workflow that wants to decouple the lifetime of the Activity from the Workflow must use a wrapper in-Workflow Activity to start a Standalone Activity using the SDK Client. In the future the Workflow will be able to start a Standalone Activity directly without a wrapper in-Workflow Activity.
